components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,243 @@
+                    val valueMap: MutableMap<String, Long> = mutableMapOf()
+                    mapData.keys().forEach { key ->
+                        valueMap[key] = mapData.tryGetLong(key) ?: 0L
+                    }
+                    valueMap
+                } catch (e: org.json.JSONException) {
+                    // This should only occur if there isn't a key/value pair stored for "values"
+                    mutableMapOf<String, Long>()
+                },
+                underflow = jsonObject.tryGetLong("underflow") ?: 0,
+                overflow = jsonObject.tryGetLong("overflow") ?: 0,
+                sum = jsonObject.tryGetLong("sum") ?: 0,
+                timeUnit = TimeUnit.values()[jsonObject.tryGetInt("timeUnit")
+                    ?: TimeUnit.Millisecond.ordinal]
+            )
+        }
+    }
+
+    // This is a helper property to build the correct identifier for the metric and allow for
+    // blank categories
+    internal val identifier: String = if (category.isEmpty()) { name } else { "$category.$name" }
+
+    // This is a list of limits for the buckets.  Instantiated lazily to ensure that the range and
+    // bucket counts are set first.
+    internal val buckets: List<Long> by lazy { getBuckets() }
+
+    /**
+     * Accumulates a sample to the correct bucket
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
+        for (i in buckets.indices) {
+            if (sample <= buckets[i] || i == bucketCount - 1) {
+                values["$i"]?.let {
+                    values["$i"] = it + 1
+                } ?: apply {
+                    values["$i"] = 1
+                }
+                break
+            }
+        }
+        sum += sample
+    }
+
+    /**
+     * Helper function to build the [TimingDistributionData] into a JSONObject for serialization
+     * purposes.
+     */
+    internal fun toJsonObject(): JSONObject {
+        return JSONObject(mapOf(
+            "category" to category,
+            "name" to name,
+            "bucketCount" to bucketCount,
+            "range" to JSONArray(range),
+            "histogramType" to histogramType.ordinal,
+            "values" to values,
+            "underflow" to underflow,
+            "overflow" to overflow,
+            "sum" to sum,
+            "timeUnit" to timeUnit.ordinal
+        ))
+    }
+
+    /**
+     * Helper function to generate the list of bucket max values used when accumulating to the
+     * correct buckets.
+     *
+     * @return List containing the bucket limits
+     */
+    private fun getBuckets(): List<Long> {
+        val logMax = Math.log(range[MAX].toDouble())
+        val result: MutableList<Long> = mutableListOf()
+        var current = range[MIN]
+        if (current == 0L) {
+            current = 1L
+        }
+        result.add(current)
+        for (i in 2..bucketCount) {
```

Comment:
Let's quicky describe the algorithm below with a comment.

Reply:
I'll try and explain it, there wasn't any documentation or good comments describing this from the desktop code where it was borrowed from either.